arrow-row/src/run.rs

```diff
@@ -0,0 +1,695 @@
+        *row = &row[consumed..];
+
+        // Check if this decoded value is the same as the previous one to identify runs
+        let is_new_run =
+            idx == 0 || decoded_data != decoded_values[*unique_row_indices.last().unwrap()];
+
+        if is_new_run {
+            // This is a new unique value - end the previous run if any
+            if idx > 0 {
+                run_ends.push(R::Native::usize_as(idx));
+            }
+            unique_row_indices.push(decoded_values.len());
+            decoded_values.push(decoded_data.clone());
+        }
+    }
+    // Add the final run end
+    run_ends.push(R::Native::usize_as(rows.len()));
+
+    // Convert the unique decoded values using the row converter
+    let mut unique_rows: Vec<&[u8]> = decoded_values.iter().map(|v| v.as_slice()).collect();
+    let values = if unique_rows.is_empty() {
+        converter.convert_raw(&mut [], validate_utf8)?
+    } else {
+        converter.convert_raw(&mut unique_rows, validate_utf8)?
+    };
+
+    // Create run ends array
+    let run_ends_array = PrimitiveArray::<R>::new(ScalarBuffer::from(run_ends), None);
+
+    // Create the RunEndEncodedArray
+    RunArray::<R>::try_new(&run_ends_array, &values[0])
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{RowConverter, SortField};
+    use arrow_array::types::Int32Type;
+    use arrow_array::{Array, Int64Array, RunArray, StringArray};
+    use arrow_schema::{DataType, SortOptions};
+    use std::sync::Arc;
+
+    #[test]
+    fn test_run_end_encoded_supports_datatype() {
+        // Test that the RowConverter correctly supports run-end encoded arrays
+        assert!(RowConverter::supports_datatype(&DataType::RunEndEncoded(
+            Arc::new(arrow_schema::Field::new("run_ends", DataType::Int32, false)),
+            Arc::new(arrow_schema::Field::new("values", DataType::Utf8, true)),
+        )));
+    }
+
+    #[test]
+    fn test_run_end_encoded_round_trip_int64s() {
+        // Test round-trip correctness for RunEndEncodedArray with Int64 values making sure it
+        // doesn't just work with eg. strings (which are all the other tests).
+
+        let values = Int64Array::from(vec![100, 200, 100, 300]);
+        let run_ends = vec![2, 3, 5, 6];
+        let array: RunArray<Int32Type> =
+            RunArray::try_new(&arrow_array::PrimitiveArray::from(run_ends), &values).unwrap();
+
+        let converter = RowConverter::new(vec![SortField::new(DataType::RunEndEncoded(
+            Arc::new(arrow_schema::Field::new("run_ends", DataType::Int32, false)),
+            Arc::new(arrow_schema::Field::new("values", DataType::Int64, true)),
+        ))])
+        .unwrap();
+
+        let rows = converter
+            .convert_columns(&[Arc::new(array.clone())])
+            .unwrap();
+
+        let arrays = converter.convert_rows(&rows).unwrap();
+        let result = arrays[0]
+            .as_any()
+            .downcast_ref::<RunArray<Int32Type>>()
+            .unwrap();
```

Comment:
You can achieve the same thing with a little less code like this if you want

```suggestion
        let result = arrays[0].as_run::<Int32Type>();
```